src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -563,27 +624,231 @@ internal SafeFileHandle OpenPort(string portName)
+            fixed (byte* bufPtr = buff)
+            {
+                // assumes dequeue-ing happens on a single thread
+                int numBytes = Interop.Sys.Read(_handle, bufPtr, buff.Length);
+
+                Interop.Error lastError = numBytes < 0 ? Interop.Sys.GetLastError() : 0;
+                if (numBytes > 0)
+                {
+                    r.Complete(numBytes);
+                    return numBytes;
+                }
+                else if (numBytes == 0)
+                {
+                    ThreadPool.QueueUserWorkItem(s => {
+                            var thisRef = (SerialStream)s;
+                            thisRef.DataReceived(thisRef, new SerialDataReceivedEventArgs(SerialData.Eof));
+                        }, this);
+                }
+                // ignore EWOULDBLOCK since we handle timeout elsewhere
+                else if (lastError != Interop.Error.EWOULDBLOCK)
+                {
+                    r.Complete(new IOException());
+                }
+            }
+
+            return 0;
+        }
+
+        private unsafe int ProcessWrite(SerialStreamAsyncResult r)
+        {
+            ReadOnlySpan<byte> buff = r.Buffer.Span;
+            fixed (byte* bufPtr = buff)
+            {
+                // assumes dequeue-ing happens on a single thread
+                int numBytes = Interop.Sys.Write(_handle, bufPtr, buff.Length);
+
+                Interop.Error lastError = numBytes < 0 ? Interop.Sys.GetLastError() : 0;
+                if (numBytes > 0)
+                {
+                    r.ProcessBytes(numBytes);
+
+                    if (r.Buffer.Length == 0)
                     {
+                        r.Complete();
                     }
-                    base.Dispose(disposing);
+
+                    return numBytes;
+                }
+                // ignore EWOULDBLOCK since we handle timeout elsewhere
+                // numBytes == 0 means that there might be an error
+                else if (lastError != Interop.Error.SUCCESS && lastError != Interop.Error.EWOULDBLOCK)
+                {
+                    r.Complete(new IOException());
+                }
+            }
+
+            return 0;
+        }
+
+        // returns number of bytes read/written
+        private static int DoIORequest(ConcurrentQueue<SerialStreamAsyncResult> q, RequestProcessor op)
+        {
+            // assumes dequeue-ing happens on a single thread
+            while (q.TryPeek(out SerialStreamAsyncResult r))
+            {
+                if (r.IsCompleted)
+                {
+                    Debug.Assert(q.TryDequeue(out _));
+                    // take another item since we haven't processed anything
+                    continue;
+                }
+
+                int ret = op(r);
+                Debug.Assert(ret >= 0);
+
+                if (r.IsCompleted)
+                {
+                    Debug.Assert(q.TryDequeue(out _));
+                }
```

Comment:
You're doing this because you want to ensure that all of the data for a write is completely written out?  If so, makes sense, but that same logic isn't needed for a read; it's fine and common for read to return less data than was asked for (as long as it returns something when the stream isn't EOF).

Reply:
This doesn't do any looping except for taking a single active request (loop is only to skip not active requests) and processing it - this doesn't wait for reading to completely finish (it could technically try to read more bytes until it reaches the timeout but it's probably matter of user preference)

Reply:
First, this seems like a really dangerous assert, as it's going to modify the queue but only in debug builds.  Is your code working in release builds?  This whole Debug.Assert expression, including the q.TryDequeue call, is going to be removed by the compiler.  Same goes for the one above.

As for looping, I was actually referring to the higher-level loop.  You're only removing the operation from the queue (Debug.Assert aside) if it's completed.  So aren't you only removing a read request once it's completed, which only happens when it's fully satisfied the requested number of bytes?

Reply:
1. I didn't know that argument evaluation can be removed - I thought only the call itself (missed this comment on my second iteration)
2. For reading I complete as soon as anything arrives, for writing it waits until completed

Reply:
> For reading I complete as soon as anything arrives, for writing it waits until completed

Ok, good. Thanks.